src/stdlib_math.fypp

```diff
@@ -1,21 +1,151 @@
+#:set RC_KINDS_TYPES = REAL_KINDS_TYPES + CMPLX_KINDS_TYPES
+#:set IRC_KINDS_TYPES = INT_KINDS_TYPES + REAL_KINDS_TYPES + CMPLX_KINDS_TYPES
 
 module stdlib_math
     use stdlib_kinds, only: int8, int16, int32, int64, sp, dp, qp
 
     implicit none
     private
-    public :: clip
+    public :: clip, linspace, logspace
+    public :: EULERS_NUMBER_SP, EULERS_NUMBER_DP, EULERS_NUMBER_QP
+
+    integer, parameter :: DEFAULT_LINSPACE_LENGTH = 100
+    integer, parameter :: DEFAULT_LOGSPACE_LENGTH = 50
+    real(dp), parameter :: DEFAULT_LOGSPACE_BASE = 10
+
+    ! Useful constants for lnspace
+    real(sp), parameter :: EULERS_NUMBER_DP = 2.71828182845904523536028747135266249775724709369995_sp
+    real(dp), parameter :: EULERS_NUMBER_SP = 2.71828182845904523536028747135266249775724709369995_dp
+    real(qp), parameter :: EULERS_NUMBER_QP = 2.71828182845904523536028747135266249775724709369995_qp
 
     interface clip
     #:for k1, t1 in IR_KINDS_TYPES
         module procedure clip_${k1}$
     #:endfor
     end interface clip
 
+    interface linspace
+    !! Version: Experimental
+    !!
+    !! Create rank 1 array of linearly spaced elements
+    !! If the number of elements is not specified, create an array with size 100. If n is a negative value,
+    !! return an array with size 0. If n = 1, return end
+    !!([Specification](../page/specs/stdlib_math.html#description))
+    #:for k1, t1 in RC_KINDS_TYPES
+      #:set RName = rname("linspace_default", 1, t1, k1)
+      module function ${RName}$(start, end) result(res)
+        ${t1}$, intent(in) :: start
+        ${t1}$, intent(in) :: end
+
+        ${t1}$ :: res(DEFAULT_LINSPACE_LENGTH)
+      end function ${RName}$
+    #:endfor
+
+    #:for k1, t1 in RC_KINDS_TYPES
+      #:set RName = rname("linspace_n", 1, t1, k1)
+      module function ${RName}$(start, end, n) result(res)
+        ${t1}$, intent(in) :: start
+        ${t1}$, intent(in) :: end
+        integer, intent(in) :: n
+
+        ${t1}$ :: res(n)
+      end function ${RName}$
+    #:endfor
+
+
+    ! Add support for integer linspace
+    !! Version: Experimental
+    !!
+    !! Create rank 1 array of linearly spaced elements from start to end.
+    !! If the number of elements is not specified, create an array with size 100. If n is a negative value,
+    !! return an array with size 0. If n = 1, return end. When dealing with integers as the `start` and `end`
+    !! parameters, the return type is always a double precision real.
+    !!
+    !!([Specification](../page/specs/stdlib_math.html#description))
+    #:for k1, t1 in INT_KINDS_TYPES
+      #:set RName = rname("linspace_default", 1, t1, k1)
+      #! The interface for INT_KINDS_TYPES cannot be combined with RC_KINDS_TYPES
+      #! because the output for integer types is always a real with dp.
+      module function ${RName}$(start, end) result(res)
+        ${t1}$, intent(in) :: start
+        ${t1}$, intent(in) :: end
+
+        real(dp) :: res(DEFAULT_LINSPACE_LENGTH)
```

Comment:
The double precision return value is used by default here.

Use the existing `fortran` generic functions to handle such functions in `fortran`: the arguments are integers, and the return values are real numbers with _different precisions_. This is very difficult. Such functions will have:

1. real(4/8/16) = zeros(integer, integer)
2. real(4/8/16) = eye(integer, integer)
3. ones
4. randu/randn
5. ...

In the future, how should we deal with these complex scenes? I think we should quickly promote a complete `fortran` generic draft.

## links

1. [https://fortran-lang.discourse.group/t/fortran-function-return-value-polymorphism/1350/10](https://fortran-lang.discourse.group/t/fortran-function-return-value-polymorphism/1350/10)
2. [https://github.com/j3-fortran/generics](https://github.com/j3-fortran/generics)
3. [https://github.com/fortran-lang/stdlib/issues/438](https://github.com/fortran-lang/stdlib/issues/438)

Reply:
The default value being double precision was a deliberate choice. If the user wants to get integer values, they can simply assign the results to an integer array:

```fortran
integer :: my_int_array(10)

my_int_array = linspace(1, 10, 10)
```

Reply:
To further elucidate my point, consider the following function call:

`linspace(1, 10)`

With a default length of the output array being 100, it is clear that the user would like the output to be a real number.

Reply:
Hmm, I understand what you mean, I'm just out of concern. Fortran's current syntax does not yet support returning a specific precision based on the precision type of the return value. This may have applications in some functions, such as
```fortran
real(4) :: x = eye(2)
real(8) :: x = eye(2)   !! Default
real(16) :: x = eye(2)
```
We can take double precision as the return value precision by default. There is no problem returning double precision to single precision, but it is not accurate to return double precision to quad precision. This is what Fortran's generics should do, and it is not perfect yet.
I am also very confused, I tried many times, and finally I gave a strange solution in [forlab/linspace](https://github.com/zoziha/forlab/blob/master/meta-src/forlab_linalg_linspace.fypp):
```fortran
call linspace(X, from, to)
call logspace(X, from, to)
call eye(X)
call zeros(X)
call ones(X)
```
I prefer functions, but I have no more options.

Reply:
I completely agree with you, and you brought up many relevant examples in your original comment. Given that nearly all modern processors run on 64-bit architectures and that quadruple precision operations can be more than 2-3 times slower than single and double precision operations, I think having double precision be the default return value is a sensible choice. 

Trust me, I hear you and I understand that this is a severe limitation of the language and the contributors of `stdlib` have to come up with a robust way to treat functions like `eye`, `zeros`, `ones`, `rand`, etc. And honestly, subroutines might just be the way to go. They avoid the copying cost from the returned array and the array that you are assigning into. They would allow use to pass integer `start` and `end` values and assign full-precision quadruple precision values to our `intent(out)` array.

But, there are also other limitations to using subroutines. Having to declare a new variable to hold the contents of `linspace`, when maybe you want to just operate on the unnamed array directly. Using subroutines to change state vs a more functional style of programming.

This PR request is clearly relevant to this discussion, but this discussion overshadows this PR and needs to be resolved in a separate issue, I believe.

Reply:
I am aware of the limitations of the language. In absence of an agreed-upon generic draft, I have *chosen* to use double precision as the default output when the input parameters are integers.

Reply:
For now, `dp` is also used as for default return values in other parts of `stdlib`. There has been some discussions about that in the start of stdlib (altough I couldn't find it back).